### ISSUE TITLE
Why interleaved in osproc.Process's output stream? and how to avoid it? 

### ISSUE THREAD

**Issue by oskca** [nim-lang/Nim#7065]
When reading the output stream of a Process started by `osproc.StartProcess` , the output is interleaved as noted in the [document](https://github.com/nim-lang/Nim/blob/a879973081e2c29d64e9fb9d8e539aa980533b10/lib/pure/osproc.nim#L87)  for `osproc.execProcess`.

I have tried using `Stream.readData` and `Stream.readStr` to read the output stream myself, but they both return the interleaved output, I don't quite understand the behavior of those standard library procedures, why are they designed to behave so wired?

**Comment by Araq** [comment]
Not sure what the problem is. Test program?

**Comment by oskca** [comment]
I have no windows machine at hand, but the program like `osprocTest.nim`:
```nim
import osproc
import streams

when defined(windows):
  let cmd = "cmd /c dir"
else:
  let cmd = "ls -lh"

echo execCmd(cmd)

let p = startProcess(cmd, options={poEvalCommand})
let s = p.outputStream()
echo s.readAll()
```
has different output in different OS (windows/wine and linux).  In windows, the output of `echo s.readAll()` is not printed, and the output of `execCmd(cmd)` is interleved with space lines

![2](https://user-images.githubusercontent.com/15141652/34860244-2d6ec238-f797-11e7-8aa6-628e4ff6f3cf.png)

**Comment by oskca** [comment]
I think the problem relates to `\r\n`
I can reproduce the problem on windows(but not linux) platform with the following codes:

`config.nims`:
```nim
import strutils
when defined(windows):
    let cmd = "cmd /c dir"
else:
    let cmd = "ls -lh"
task bad, "bad output":
    echo staticExec(cmd).replace("\n", "\r\n")
task good, "bad output":
    echo staticExec(cmd)
```

`osprocTest.nim`:
```nim
import osproc
import streams
let goodcmd = "nim good"
let badcmd = "nim bad"
echo execCmd(goodcmd)
echo execCmd(badcmd)
```

Though `nim good` and `nim bad` all produce the same output (by sight), when you run `nim c -r osprocTest.nim` you can see interleaved output for `echo execCmd(badcmd)` on the console.

This is perhaps not an serious issue, but why should procs like `osproc.execCmd` or  `osproc.startProcess via outputStream()`  alter the true output?

**Comment by cheatfate** [comment]
@oskca in this sample
```nim
import osproc
import streams

when defined(windows):
  let cmd = "cmd /c dir"
else:
  let cmd = "ls -lh"

echo execCmd(cmd)

let p = startProcess(cmd, options={poEvalCommand})
let s = p.outputStream()
## add here `discard waitForExit(p)`
echo s.readAll()
```
you need to wait for child process exit... your sample is exiting before child process exited, so you have read not all stdout of child process.

**Comment by oskca** [comment]
@cheatfate yes, `execProcess` use `waitForExit(p)` as a signal to indicate the running state of the target process `p`, but as for `p.outputStream()` the return value is a new object of type `streams.Stream`, new users would depends on this new stream until they encounter the problem and then look into to stdlib to find that they have to use `waitForExit(p)` to read the full output. **This behavior** is not documented in the stdlib docs, and is really frustrated to finally find out an usable solution after a long time probing around.

I like `NIM` very much personally, but this kind of thing is really a pain in the ass for normal `NIM` users, especially compared to `Go` and it's documentations.

**Comment by cheatfate** [comment]
@dom96, please remove `stdlib` label from this issue.

**Comment by cheatfate** [comment]
@oskca this project is maintained by 10 people (most of which are not native english speakers), but not by multi-billionaire corporation with tens thousands of workers and contributors, so yeah our documentation is not so brilliant and shiny.

**Comment by dom96** [comment]
> @dom96, please remove stdlib label from this issue.

Why? It refers to a stdlib module.

**Comment by oskca** [comment]
yes, the interleaved problem is still there

**Comment by oskca** [comment]
@cheatfate yes, your people are all great and nice folks, I appreciate all your work very much.This problem itself is not a big deal, but it's an important issue when `Nim` as a programming language is introduced to the bigger audience. 

BTW: I'm using `Nim` in my personal projects recently, though I don't have an deep knowledge of the whole `Nim` ecosystem, I like it very much, and I'm trying to mastering it.
